packages/astro/src/config/vite-plugin-content-listen.ts

```diff
@@ -0,0 +1,41 @@
+import type fsMod from 'node:fs';
+import type { Plugin, ViteDevServer } from 'vite';
+import type { AstroSettings } from '../@types/astro';
+import { attachContentServerListeners } from '../content/server-listeners.js';
+import type { LogOptions } from '../core/logger/core.js';
+
+/**
+ * Listen for Astro content directory changes and generate types.
+ *
+ * This is a separate plugin for `getViteConfig` as the `attachContentServerListeners` API
+ * needs to be called at different times in `astro dev` and `getViteConfig`. For `astro dev`,
+ * it needs to be called after the Astro server is started (packages/astro/src/core/dev/dev.ts).
+ * For `getViteConfig`, it needs to be called after the Vite server is started.
+ */
+export function astroContentListenPlugin({
+	settings,
+	logging,
+	fs,
+}: {
+	settings: AstroSettings;
+	logging: LogOptions;
+	fs: typeof fsMod;
+}): Plugin {
+	let server: ViteDevServer;
+
+	return {
+		name: 'astro:content-listen',
+		apply: 'serve',
+		configureServer(_server) {
+			server = _server;
+		},
+		async buildStart() {
+			await attachContentServerListeners({
+				fs: fs,
+				settings,
+				logging,
+				viteServer: server,
+			});
+		},
+	};
```

Comment:
A bit confused when this runs. Is this potentially triggering the listeners _twice_ (once from the `dev.ts` file and once from here)? Or is this exclusively a patch for vitest?

Reply:
The `dev.ts` file would only run when we call `astro dev`, so this call here is exclusively for vitest, since a vitest/vite config setup doesn't call the Astro CLI.

Reply:
Got it. Another reason to remove the server requirement for loading content configs! Can ideally use a vite plugin for all cases instead of calling `attach-listeners` separately 👍

Reply:
I did tried a Vite plugin at first, but the timing when it logs "content types generated" for the CLI is off (logs before the CLI), and it's pretty hard to adjust that timing that I went with calling it in two places for now 😬